components/org.wso2.carbon.editor.log.appender/src/main/java/org/wso2/carbon/editor/log/appender/internal/CircularBuffer.java

```diff
@@ -0,0 +1,167 @@
+     */
+    public CircularBuffer(int size) {
+        if (size <= 0) {
+            throw new IllegalArgumentException(
+                    "Requested size of circular buffer (" + size + ") is invalid");
+        }
+        if (size > MAX_ALLOWED_SIZE) {
+            throw new IllegalArgumentException(
+                    "Requested size of circular buffer (" + size + ") is greater than the " +
+                            "allowed max size " + MAX_ALLOWED_SIZE);
+        }
+        this.size = size;
+        this.bufferList = new ArrayList<E>(getSize());
+        this.startIndex = 0;
+        this.endIndex = -1;
+    }
+
+    /**
+     * Create a circular buffer with the maximum allowed size
+     */
+    public CircularBuffer() {
+        this(MAX_ALLOWED_SIZE);
+    }
+
+    /**
+     * Append elements while preserving the circular nature of the buffer.
+     *
+     * @param element - element to be appended
+     */
+    public synchronized void append(E element) {
+        if (element == null) {
+            throw new IllegalArgumentException(
+                    "Circular buffer doesn't support null values to be added to buffer");
+        }
+        if (startIndex == getSize() - 1) {
+            startIndex = 0;
+        } else if (endIndex == getSize() - 1) {
+            endIndex = -1;
+            startIndex = 1;
+        } else if (startIndex != 0) {
+            // start index is not in beginning of the buffer
+            startIndex++;
+        }
+        endIndex++;
+        if (getSize() == bufferList.size()) {
+            // if the buffer capacity has been reached, replace the existing elements,
+            // set method replaces the element in the given index
+            bufferList.set(endIndex, element);
+        } else {
+            // if the buffer capacity has not been reached add elements to the list,
+            // add method lets the array list grow, and appends elements to the end of list
+            bufferList.add(endIndex, element);
+        }
+    }
+
+    /**
+     * Retrieve the given amount of elements from the circular buffer. This is a forgiving
+     * operation, if the amount asked is greater than the size of the buffer it will return all the
+     * available elements
+     *
+     * @param amount - no of elements to return
+     * @return - a list of elements
+     */
+    public synchronized List<E> get(int amount) {
+        if (bufferList.isEmpty()) {
+            // if the container is empty return an empty list
+            return new ArrayList<E>();
+        }
+        if (amount <= 0) {
+            // if a negative amount is requested send an empty list
+            return new ArrayList<E>();
+        }
+        int amountOfElementsInContainer = bufferList.size();
+        int amountToRetrieve = amount;
+        List<E> result = new ArrayList<E>(amountOfElementsInContainer);
+        /*for (int i = startIndex; amountOfElementsInContainer > 0 && amountToRetrieve > 0;
```

Comment:
I think we can remove this comment

Reply:
Updated the method in [#68464573e9cd97eb8ba4b74b2baa4859a359e17d](https://github.com/wso2/carbon-analytics/commit/68464573e9cd97eb8ba4b74b2baa4859a359e17d)